indexer/feature_altitude.hpp

```diff
@@ -0,0 +1,124 @@
+
+namespace feature
+{
+using TAltitude = int16_t;
+using TAltitudes = vector<feature::TAltitude>;
+using TAltitudeSectionOffset = uint32_t;
+
+TAltitude constexpr kInvalidAltitude = numeric_limits<TAltitude>::min();
+
+struct AltitudeHeader
+{
+  using TAltitudeSectionVersion = uint16_t;
+
+  AltitudeHeader()
+  {
+    Reset();
+  }
+
+  template <class TSink>
+  void Serialize(TSink & sink) const
+  {
+    sink.Write(&version, sizeof(version));
+    sink.Write(&minAltitude, sizeof(minAltitude));
+    sink.Write(&featureTableOffset, sizeof(featureTableOffset));
+    sink.Write(&altitudeInfoOffset, sizeof(altitudeInfoOffset));
+    sink.Write(&endOffset, sizeof(endOffset));
+  }
+
+  template <class TSource>
+  void Deserialize(TSource & src)
+  {
+    src.Read(&version, sizeof(version));
+    src.Read(&minAltitude, sizeof(minAltitude));
+    src.Read(&featureTableOffset, sizeof(featureTableOffset));
+    src.Read(&altitudeInfoOffset, sizeof(altitudeInfoOffset));
+    src.Read(&endOffset, sizeof(endOffset));
+  }
+
+  // Methods below return sizes of parts of altitude section in bytes.
+  static size_t GetHeaderSize() { return sizeof(AltitudeHeader); }
+
+  size_t GetAltitudeAvailabilitySize() const { return featureTableOffset - GetHeaderSize(); }
+  size_t GetFeatureTableSize() const { return altitudeInfoOffset - featureTableOffset; }
+  size_t GetAltitudeInfo() const { return endOffset - altitudeInfoOffset; }
+
+  void Reset()
+  {
+    version = 1;
+    minAltitude = kInvalidAltitude;
+    featureTableOffset = 0;
+    altitudeInfoOffset = 0;
+    endOffset = 0;
+  }
+
+  TAltitudeSectionVersion version;
+  TAltitude minAltitude;
+  TAltitudeSectionOffset featureTableOffset;
+  TAltitudeSectionOffset altitudeInfoOffset;
+  TAltitudeSectionOffset endOffset;
+};
+
+class Altitude
+{
+public:
+  Altitude() = default;
+  explicit Altitude(TAltitudes const & altitudes) : m_altitudes(altitudes) {}
+
+  template <class TSink>
+  void Serialize(TAltitude minAltitude, TSink & sink) const
+  {
+    CHECK(!m_altitudes.empty(), ());
+
+    WriteVarInt(sink, static_cast<int32_t>(m_altitudes[0] - static_cast<int32_t>(minAltitude)));
+    for (size_t i = 1; i < m_altitudes.size(); ++i)
+      WriteVarInt(sink, static_cast<int32_t>(m_altitudes[i]) - static_cast<int32_t>(m_altitudes[i - 1]));
+  }
+
+  template <class TSource>
+  void Deserialize(TAltitude minAltitude, size_t pointCount, TSource & src)
+  {
+    m_altitudes.clear();
+    if (pointCount == 0)
+    {
+      ASSERT(false, ());
+      return;
+    }
+
+    m_altitudes.resize(pointCount);
+    TAltitude prevPntAltitude = minAltitude;
+    for (size_t i = 0; i < pointCount; ++i)
+    {
+      m_altitudes[i] = static_cast<TAltitude>(ReadVarInt<int32_t>(src) + prevPntAltitude);
+      if (m_altitudes[i] < minAltitude)
+      {
+        ASSERT(false, ());
+        m_altitudes.clear();
+        return;
+      }
+      prevPntAltitude = m_altitudes[i];
+    }
+  }
+
+  TAltitudes GetAltitudes() const
```

Comment:
I guess it's possible. Then it'll be used in the following code:
```c++
map<uint32_t, TAltitudes> m_cache;
.....
  {
    Altitudes a;
.....
    return m_cache.insert(make_pair(featureId, a.GetAltitudes())).first->second;
  }
```
I guess even if TAltitudes is returned by const ref, make_pair creates a pair<uint32_t, TAltitudes> and while creating the pair Altitudes will by copied.

Now. I'll return TAltitudes by cosnt ref and use it in the following way.
```
return m_cache.insert(make_pair<uint32_t, TAltitudes>(featureId, a.GetAltitudes())).first->second;
```